return 503 for aggregated APIs when the APIServiceRegistrationController hasn't finished installing all known APIServices

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug


<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it: 
Ensures that requests to aggregated APIs will get 503, not 404 while the APIServiceRegistrationController hasn't finished installing all known APIServices. It is especially important for controllers like GC and NS since they act on 404

Note that this is an already existing behavior (returning 503 when the CRD informer hasn't been sync). 
We already have it for the apiextentionserver, see https://github.com/kubernetes/kubernetes/pull/105653 for more details.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #104342

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
resolves a potential issue with GC and NS controllers which may delete objects after getting a 404 response from the server during its startup. This PR ensures that requests to aggregated APIs will get 503, not 404 while the APIServiceRegistrationController hasn't finished its job.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

p0lyn0mial committed:
allow for passing a custom handler to the empty delegate

p0lyn0mial commented:
/assing @tkashem @sttts

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -22,6 +22,7 @@ package app
 import (
 	"crypto/tls"
 	"fmt"
+	"k8s.io/apiserver/pkg/util/notfoundhandler"
```
order

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -271,14 +271,19 @@ func (s *GenericAPIServer) NextDelegate() DelegationTarget {
+
 func (s emptyDelegate) UnprotectedHandler() http.Handler {
-	return nil
+	return s.handler
```
remind me why we only need this for the unprotected handler

p0lyn0mial replied:
this is the one we use at the end of the chain, here https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L190

MikeSpreitzer replied:
Can you please help me understand, by spelling out the rest of the answer?

sttts commented on staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go:
```diff
@@ -0,0 +1,72 @@
+	"k8s.io/apiserver/pkg/util/signal"
+)
+
+// New returns an HTTP handler that is meant to be executed at the end of the delegation chain and check if the server has been fully prepared before replying to the client.
```
mention that this is long before readiness and is meant to block request before the muxes and handlers are completely initialized such that requests are properly routed.

sttts replied:
And that preparedToServe must not depend on other components in the cluster other than etcd

sttts commented on staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go:
```diff
@@ -0,0 +1,72 @@
+
+func (h *Handler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	for _, preparedToServeSignal := range h.preparedToServeSignals {
+		if !preparedToServeSignal.IsSignaled() {
```
continue early

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -113,10 +114,26 @@ type preparedAPIAggregator struct {
 }
 
+// preparedToServeSignals holds signals/events that indicate this server has been fully initialized and is ready to serve requests
+type preparedToServeSignals struct {
```
why a wrapper around another type? Just use `signal.Interface` directly

p0lyn0mial replied:
I wanted to expose only the read-only interface

MikeSpreitzer replied:
But the current text exposes the full interface.
It looks to me like the reason for a struct here is to prepare for more signals in the future.  For example, one about RBAC being initialized.

p0lyn0mial commented:
I will remove the signal pkg and use raw types instead. I will also add a signal from the extension API server.

tkashem commented on staging/src/k8s.io/apiserver/pkg/util/signal/signal.go:
```diff
@@ -0,0 +1,63 @@
+	case <-e.ch:
+		// already closed, don't close again.
+	default:
+		close(e.ch)
```
if `Signal` is used  concurrently, i think we need to use a `sync.Once` to protect it.

tkashem commented on staging/src/k8s.io/apiserver/pkg/util/signal/signal.go:
```diff
@@ -0,0 +1,63 @@
+	}
+}
+
+type namedChannelWrapper struct {
```
since we are converting this to a shared utility, i would recommend unit tests for it.

caesarxuchao commented:
/cc @cheftako @jpbetz 

Walter, I think you mentioned a specific http status code during the sig meeting.

caesarxuchao commented:
/triage accepted

tkashem commented:
cc @MikeSpreitzer

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -271,14 +271,19 @@ func (s *GenericAPIServer) NextDelegate() DelegationTarget {
 }
 
 type emptyDelegate struct {
+	handler http.Handler
```
What did "empty delegate" mean, exactly?  Is it still empty?

MikeSpreitzer commented:
We discussed this in the Aug 25 SIG api-machinery meeting.  If I recall correctly, the consensus was that the client should be made to opt in to this.  We have started drafting a KEP, at https://docs.google.com/document/d/1JQNT0eu6Gpn9UZ4rcfIUDkJf1wKc3f_57KKC4c-_s-U .

MikeSpreitzer commented:
Also, we want this conditionality to cover RBAC initialization too.  A too-early answer in that regard is an authorization failure rather than a 404.

cheftako commented:
> /cc @cheftako @jpbetz
> 
> Walter, I think you mentioned a specific http status code during the sig meeting.

Partly depends on the idea. 
I liked the precondition solution in which case I think 412 would be the correct response code.
It seemed a more backward compatible fix in that clients would clearly be opting in.

However this seems to be a general "KAS not ready yet" solution.
For that I think 503 is the correct response node.

p0lyn0mial committed:
update modules.txt

p0lyn0mial committed:
apiserver: adds WithMuxCompleteProtection filter

It puts the muxCompleteProtectionKey in the context if a request has been made before muxCompleteSignal has been ready.
Putting the key protect us from returning a 404 response instead of a 503.
It is especially important for controllers like GC and NS since they act on 404s.

The presence of the key is checked in the NotFoundHandler (staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go)

The race may happen when a request reaches the NotFoundHandler because not all paths have been registered in the mux
but when the registered checks are examined in the handler they indicate that the paths have been actually installed.
In that case, the presence of the key will make the handler return 503 instead of 404.

p0lyn0mial committed:
kube-aggregator: registeres APIServiceRegistrationControllerInitiated as a MuxComplete signal

apiServiceRegistrationControllerInitiated is closed when APIServiceRegistrationController has finished "installing" all known APIServices.
At this point we know that the proxy handler knows about APIServices and can handle client requests.
Before it might have resulted in a 404 response which could have serious consequences for some controllers like GC and NS

p0lyn0mial committed:
kube-apiserver: wires the notFoundHandler

p0lyn0mial committed:
apiserver: indroduces NotFoundHanlder

The new handler is meant to be executed at the end of the delegation chain.
It simply checks if the request have been made before the server has installed all known HTTP paths.
In that case it returns a 503 response otherwise it returns a 404.

We don't want to add additional checks to the readyz path as it might prevent fixing bricked clusters.
This specific handler is meant to "protect" requests that arrive before the paths and handlers are fully initialized.

p0lyn0mial committed:
genericapiserver: indroduce muxCompleteSignals for holding signals that indicate all known HTTP paths have been registered

the new field exists primarily to avoid returning a 404 response when a resource actually exists but we haven't installed the path to a handler.
it is exposed for easier composition of the individual servers.
the primary users of this field are the WithMuxCompleteProtection filter and the NotFoundHandler.

p0lyn0mial commented:
/retest

p0lyn0mial committed:
rename to muxAndDiscoveryComplete

sttts commented:
Lgtm and approved.

Will give @MikeSpreitzer @caesarxuchao and others a chance to take a final look.

MikeSpreitzer reviewed (commented):
What about https://docs.google.com/document/d/1JQNT0eu6Gpn9UZ4rcfIUDkJf1wKc3f_57KKC4c-_s-U ?  My recollection of the SIG api-machinery discussion is that a client opt-in was favored, out of concern that a non-optional solution would ruining some startup sequences that involve multiple processes.

p0lyn0mial committed:
aggregator: pass apiServiceRegistrationControllerInitiated signal directly to apiserviceRegistration controller

p0lyn0mial commented:
@sttts @MikeSpreitzer many thanks for reviewing the PR. I applied the last batch of comments. I think we are ready to tag.

p0lyn0mial commented:
/test pull-kubernetes-unit

MikeSpreitzer reviewed (commented):
This looks basically OK to me.
I have just a few minor comments.
